adding ability to reconfigure an active cluster.

albertshau opened the pull request.

albertshau committed:
adding ability to reconfigure an active cluster.

rns4731 committed:
cluster reconfig form

rns4731 committed:
adding json modifier

rns4731 committed:
json validation added

rns4731 committed:
add ability to add services

rns4731 committed:
adding reconfigure-cluster to this branch

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/Actions.java:
```diff
@@ -115,7 +134,14 @@ public Actions() {
+                      ImmutableSet.of(
+                        // Start of a dependent service can happen only after its dependency has started
+                        new Dependency(ProvisionerAction.START, ProvisionerAction.START),
+                        // Stop of a dependent service can happen only after its dependency has started
```
"after its dependency has started" -> "after its dependency has stopped"

albertshau replied:
done.  It should actually be "before its dependency has stopped" too.

poornachandra commented on server/src/main/java/com/continuuity/loom/scheduler/task/ClusterService.java:
```diff
@@ -39,6 +41,8 @@
  */
 public class ClusterService {
   private static final Logger LOG = LoggerFactory.getLogger(ClusterService.class);
+  private static final ImmutableSet<Cluster.Status> CONFIGURABLE_STATES =
```
Maybe move this to the file where Cluster Status is defined.

chtyim replied:
Use more generic interface, such as `Set`, rather than a specific implementation as type.

albertshau replied:
done for both

poornachandra commented on server/src/main/java/com/continuuity/loom/store/SQLClusterStore.java:
```diff
@@ -177,10 +177,37 @@ public Cluster getCluster(String clusterId, String ownerId) throws Exception {
     // sticking with standard sql... this could be done in one step with replace, or with
     // insert ... on duplicate key update with mysql.
-    long clusterNum = Long.valueOf(cluster.getId());
+    long clusterNum = Long.parseLong(cluster.getId());
```
Just curious - why is valueOf changed to parseLong?

albertshau replied:
feedback from a previous PR, parseLong returns a long whereas valueOf returns the Long object so better to use parseLong since the left is a long.

poornachandra commented:
Minor comments, otherwise LGTM.

I haven't reviewed the html/css/js files though. I assume they are result of a merge.

chtyim commented on server/src/main/java/com/continuuity/loom/store/SQLClusterStore.java:
```diff
@@ -714,6 +741,26 @@ public void deleteNode(String nodeId) throws Exception {
+        rs.close();
+      }
+    } finally {
+      statement.close();
```
Not a good practice to close a resource being passed to a method (except a util method for performing the close, e.g. in `Closeables.closeQuietly()`). Resource acquisition and release should be in the same block (in this case, caller should be closing the statement).

albertshau replied:
done

albertshau committed:
close resources in SQLClusterStore where they are created,
cleaning up a comment, and moving the set of configurable states
to the cluster status.

albertshau committed:
Merge branch 'feature/reconfigure-cluster' of github.com:continuuity/loom into feature/reconfigure-cluster
